instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,679 @@
+        recurse_namedtuple_attribs (dict): A dictionary of ``BaseEstimator``-
+          derived sklearn types as keys, with values being a list of
+          attribute names. Each attribute represents a list of 2-tuples in
+          which the first element is the estimator name, and the second
+          element is the estimator. Defaults include sklearn's ``Pipeline``
+          and its attribute ``steps``, and the ``FeatureUnion`` and its
+          attribute ``transformer_list``. Used in model instrumentation only.
+        spanner: A function with signature (func, str) which
+          decorates instance methods with opentelemetry spans.
+        packages: A list of additional sklearn-compatible packages to
+          instrument. Used with package instrumentation only.
+        exclude_classes: A list of classes to exclude from instrumentation.
+          Child classes are also excluded. Default is sklearn's
+          ``[BaseDecisionTree]``.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """Override new.
+
+        The base class' new method passes args and kwargs. We override because
+        we init the class with configuration and Python raises TypeError when
+        additional arguments are passed to the object.__new__() method.
+        """
+        if cls._instance is None:
+            cls._instance = object.__new__(cls)
+
+        return cls._instance
+
+    def __init__(
+        self,
+        methods: List[str] = None,
+        recurse_attribs: Dict[Type[BaseEstimator], List[str]] = None,
+        recurse_namedtuple_attribs: Dict[
+            Type[BaseEstimator], List[str]
+        ] = None,
+        spanner: Callable = implement_spans,
+        packages: List[str] = None,
+        exclude_classes: List[Type] = None,
+    ):
+        self.methods = methods or DEFAULT_METHODS
+        self.recurse_attribs = recurse_attribs or DEFAULT_ATTRIBS
+        self.recurse_namedtuple_attribs = (
+            recurse_namedtuple_attribs or DEFAULT_NAMEDTUPLE_ATTRIBS
+        )
+        self.spanner = spanner
+        self.packages = packages or DEFAULT_PACKAGES
+        if exclude_classes is None:
+            self.exclude_classes = tuple(DEFAULT_EXCLUDE_CLASSES)
+        else:
+            self.exclude_classes = tuple(exclude_classes)
+
+    def _instrument(self, **kwargs):
+        """Instrument the library, and any additional specified on init."""
+        klasses = get_base_estimators(packages=self.packages)
+        for _, klass in klasses.items():
+            if issubclass(klass, self.exclude_classes):
+                logger.debug("Not instrumenting (excluded): %s", str(klass))
+            else:
+                logger.debug("Instrumenting: %s", str(klass))
+                for method_name in self.methods:
+                    if hasattr(klass, method_name):
+                        self._instrument_class_method(
+                            estimator=klass, method_name=method_name
```

Comment:
Just a concern, what if there is another instrumentation also installed for the same package that is passed as an argument for this instrumentor? Would that cause double instrumentation?

Reply:
The only way it wouldn't is if that other instrumentation package used the same strategy, i.e. resulted in a True return value from the `_check_instrumented` method. Order of instrumentation might also matter in that regard.

Related, one of the things I tried to do here was abstract the span decorator as an instrumentor arg with the idea that multiple instrumentations (say otel + datadog) could be applied to estimators by just instantiating multiple instrumentors with different decorators. However, this isn't feasible with the current code because of how instrumentation is applied with the `_original_xxx` attributes and the `_check_instrumented` method.